### ISSUE TITLE
Improve performance of `np.arange(Time, Time, TimeDelta)`

### ISSUE THREAD

**Issue by dstansby** [astropy/astropy#12700]
### Description
Creating a regularly space set of `Time` objects with `np.linspace` is quite slow:
```python
from astropy.time import Time, TimeDelta
import astropy.units as u
import numpy as np

dt = TimeDelta(1 * u.hour)
t0 = Time('2021-01-01')
t1 = Time('2022-01-01')
```
```python
%timeit Time(np.arange(t0, t1, dt))
3.13 s ± 443 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
Manually converting to a float and calling arange on that is much faster:
```python
In [10]: %timeit Time(np.arange(t0.jd, t1.jd, dt.to_value(u.day)), format='jd')
360 µs ± 9.66 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
It would be nice to improve performance of the first option, so one doesn't have to convert back and forth to floats to get decent performance.

**Comment by mkbrewer** [comment]
This is likely due to having to instantiate all those `Time()` objects. As an alternative, you might try:
```
from datetime import datetime, timedelta
t0 = datetime(2021, 1, 1)
t1 = datetime(2022, 1, 1)
dt = timedelta(hours = 1)

**Comment by mhvk** [comment]
Yes, it should be possible to do this, by giving `Time` an `__array_function__` implementation - see the overall issue #8610 (where `np.linspace` is mentioned, though I hadn't realized it sort-of worked already!). Sadly, I've not gotten around to actually implementing this... But willing to help!

**Comment by dstansby** [comment]
👍  thanks for the pointer in that direction - I'll have a go at implementing `__array_funciton__` for `Time`.

**Comment by dstansby** [comment]
From https://github.com/numpy/numpy/issues/12379 it looks like `arange` isn't supported yet with `__array_function__` on the numpy side.

**Comment by mhvk** [comment]
Hmm, true, but `np.linspace` is supported, and I think that is the more relevant function (which you also listed on top, though your example is with `np.arange`).

For `np.arange`, there is no direct support since by default the arguments are not arrays, so there is no reason to redirect on those. But in principle one can make it pass through `__array_function__` by passing in `like=Time`. Not sure that is worth it, though... (I haven't done it for `Quantity`)

**Comment by dstansby** [comment]
I've just got `np.linspace` working. Going to add a (long!) list of functions that don't make sense for times, and then open a PR.

**Comment by mhvk** [comment]
Sounds great :fireworks: 

For any function that you are not sure, you can just fall back to the default implementation. Easiest for that is probably just to fall back to `super().__array_function__(...)` -- this gets you to `astropy.utils.shapes.ShapedNDArray.__array_function__`, which already deals with some of the functions.

**Comment by taldcroft** [comment]
> But in principle one can make it pass through __array_function__ by passing in like=Time. Not sure that is worth it, though... (I haven't done it for Quantity)

I don't quite understand why the `like=Time` would be needed, I thought that `__array_function__` would try to delegate to the `Time` class if the first arg is a `Time`.

In any case, it definitely seems to me that `np.arange(t0, t1, dt)` should return a single `Time` object if `t0` is a `Time`, instead of an array of `Time` objects.

**Comment by mhvk** [comment]
@taldcroft - it is just how numpy has done the function override - you can only override arguments that would in normal numpy use take an array as input. For `np.arange`, the regular arguments are always scalars and hence they cannot be overridden on the argument. It is a similar situation for `np.zeros`, where none of the arguments are arrays (and which also has a `like=` argument to give the class of the result you would like). 

Thankfully, I had previously adjusted `np.linspace` to take arrays for start and end (for `np.linspace` this makes conceptual sense, since the number of points is always the same, unlike for `np.arange`). Hence, `np.linspace` can work with start and stop being `Time` instances without any problem.

**Comment by taldcroft** [comment]
In the meantime using `np.arange` seems to generate a warning from `TimeDelta` despite giving the expected result:
```
In [16]: t1 = Time("2020-01-01")

In [17]: t2 = Time("2020-01-02")

In [18]: np.arange(t1, t2, 6 * u.hr)
WARNING: TimeDeltaMissingUnitWarning: Numerical value without unit or explicit format passed to TimeDelta, assuming days [astropy.time.core]
Out[18]: 
array([<Time object: scale='utc' format='iso' value=2020-01-01 00:00:00.000>,
       <Time object: scale='utc' format='iso' value=2020-01-01 06:00:00.000>,
       <Time object: scale='utc' format='iso' value=2020-01-01 12:00:00.000>,
       <Time object: scale='utc' format='iso' value=2020-01-01 18:00:00.000>],
      dtype=object)
```
I'm not sure where that warning comes from but it would be nice if this worked without a warning. (Accepting the slow performance and object array of Time objects).

**Comment by neutrinoceros** [comment]
I'm not sure the result is quite what one would expect: while the values are numerically correct, you don't get a single `Time` instance, but a numpy array with dtype=object.

As pointed out previously, I believe the proper way to support this use case would be to 
- require that the `like` keyword argument be passed
- implement an `__array_function__` dispatch for `np.arange` either for the `Time` class or one of its parents, as suited

Note that I have a PR open to do the same thing for `units.Quantity` (and this specific function is proving to be a challenge, but it's manageable), see #17059
